Improving Documentation (#973)
iotaledger/hornet: docs/post_installation/post_installation.md

# Post-installation steps

> Once the Hornet is deployed, all parameters are set via configuration files.

Reviewer: ```suggestion
Once Hornet is deployed, all parameters are set via configuration files.
```

## Configuration
The most important ones are:

> * `config.json`: include all configuration flags and their values

Reviewer: ```suggestion
* `config.json`: includes all configuration flags and their values
```

> * `peering.json`: include all connection details to your static peers (neighbors)

Reviewer: ```suggestion
* `peering.json`: includes all connection details to your static peers (neighbors)
```

> Since the Hornet node software is able to power original IOTA 1.0 network as well as IOTA 1.5 (aka Chrysalis), it is important to use respective `config.json` file that targets the IOTA network that we want. All configuration files that targets respective networks are available in [source code repo](https://github.com/gohornet/hornet/tree/master) on GitHub.

Reviewer: That's not correct, there are distinct versions of Hornet which are compatible to the different protocols but there isn't a version which supports all. The configuration for these versions are completely different.

Author: Great catch. Thank you.

Author: Changed

> Depending on installation path you selected, default config files may be also a part of the installation experience and so you may see the following configuration files at your deployment directory:

Reviewer: ```suggestion
Depending on the installation path you selected, default config files may be also be part of the installation process and so you may see the following configuration files at your deployment directory:
```

```bash
config.json
config_chrysalis_testnet.json
config_comnet.json
config_devnet.json
peering.json
profiles.json
```

### Default configuration
By default, Hornet searches for configuration files in a working directory and expects default names, such as `config.json` and `peering.json`.

This behavior can be changed by running Hornet with some altering arguments. The default directory with configuration files can be changed by running `hornet --config-dir` argument.

Please see [config.json](./config.md) and [peering.json](./peering.md) chapters for more information regarding respective configuration files.

> Once Hornet is executed, it outputs all loaded configuration parameters to `stdout` to be sure what configuration was loaded.

Reviewer: ```suggestion
Once Hornet is executed, it outputs all loaded configuration parameters to `stdout` to show what configuration was loaded.
```

All other altering command line parameters can be obtained by running `hornet --help`.

> Hornet version 0.5.x targets IOTA 1.0 mainnet network by default. Hornet version 0.6.x targets IOTA 1.5 (Chrysalis) mainnet network by default

### Identifying a configuration file based on particular use cases
There is a simple way how to recognize which configuration file targets which IOTA network:
```bash
cat config.json | jq "[.httpAPI?, .restAPI?]"
```
* `jq` command line json parser can be installed using `sudo apt install jq`

IOTA 1.5 (Chrysalis network) provides the following rest API endpoints:
```json

> [

Reviewer: this shouldn't be an array

  null,
  {
    "jwtAuth": {
      "enabled": false,
      "salt": "HORNET"
    },
    "excludeHealthCheckFromAuth": false,
    "permittedRoutes": [
      "/health",
      "/mqtt",
      "/api/v1/info",
      "/api/v1/tips",
      "/api/v1/messages/:messageID",
      "/api/v1/messages/:messageID/metadata",
      "/api/v1/messages/:messageID/raw",
      "/api/v1/messages/:messageID/children",
      "/api/v1/messages",
      "/api/v1/transactions/:transactionID/included-message",
      "/api/v1/milestones/:milestoneIndex",
      "/api/v1/milestones/:milestoneIndex/utxo-changes",
      "/api/v1/outputs/:outputID",
      "/api/v1/addresses/:address",
      "/api/v1/addresses/:address/outputs",
      "/api/v1/addresses/ed25519/:address",
      "/api/v1/addresses/ed25519/:address/outputs",
      "/api/v1/treasury"
    ],
    "whitelistedAddresses": [
      "127.0.0.1",
      "::1"
    ],
    "bindAddress": "0.0.0.0:14265",
    "powEnabled": true,
    "powWorkerCount": 1,
    "limits": {
      "bodyLength": "1M",
      "maxResults": 1000
    }
  }
]
```

> IOTA 1.0 provides legacy rest API endpoints:

Reviewer: No point in including instructions for IOTA 1.0 here.

Author: Removed

```json
[
  {
    "basicAuth": {
      "enabled": false,
      "username": "",
      "passwordHash": "",
      "passwordSalt": ""
    },
    "excludeHealthCheckFromAuth": false,
    "permitRemoteAccess": [
      "getNodeInfo",
      "getBalances",
      "checkConsistency",
      "getTipInfo",
      "getTransactionsToApprove",
      "getInclusionStates",
      "getNodeAPIConfiguration",
      "wereAddressesSpentFrom",
      "broadcastTransactions",
      "findTransactions",
      "storeTransactions",
      "getTrytes"
    ],
    "permittedRoutes": [
      "healthz"
    ],
    "whitelistedAddresses": [],
    "bindAddress": "0.0.0.0:14265",
    "limits": {
      "bodyLengthBytes": 1000000,
      "findTransactions": 1000,
      "getTrytes": 1000,
      "requestsList": 1000
    }
  },
  null
]
```

## Dashboard
There is an admin dashboard available in Hornet (on port 8081) and it is enabled by default. It provides some useful information regarding the node, its internal activity and its resources, such as memory, peers, transactions per second rate, etc.

> However it is not listening to incoming requests from a public traffic to prevent access from a malicious actor. It is listening only to requests from `localhost` by default.

Reviewer: "from a public traffic" is not proper English.

Author: Changed

It can be configured via the following `config.json` file section:

```json
"dashboard": {
    "bindAddress": "localhost:8081",
    "auth": {
      "sessionTimeout": "72h",
      "username": "admin",
      "passwordHash": "0000000000000000000000000000000000000000000000000000000000000000",
      "passwordSalt": "0000000000000000000000000000000000000000000000000000000000000000"
    }
  }
```
* to enable Dashboard to be reachable from a public traffic, it can be changed to `"bindAddress": "0.0.0.0:8081"`

> Even if enabled to public traffic, a visitor still needs a valid combination of username and password to access a management part of the dashboard.

Reviewer: again, "to public traffic" is not correct wording here.

It can be generated using integrated command-line Hornet tools, specifically `pwdhash`:
```bash
hornet tools pwdhash
```

Output example:
```plaintext
Enter a password:
Re-enter your password:
Success!
Your hash: 24c832e35dc542901b90888321dbfc4b1d9617332cbc124709204e6edf7e49f9
Your salt: 6c71f4753f6fb52d7a4bb5471281400c8fef760533f0589026a0e646bc03acd4
```
* `pwdhash` tool provides a combination of `passwordHash` and `passwordSalt` based on your input password
* both values should be copied to the respective configuration values above

> Once Hornet is restarted then the Dashboard is accessible using the given credentials.

Reviewer: ```suggestion
Once Hornet is restarted then the dashboard is accessible using the given credentials.
```

# Post-installation
Once Hornet is deployed, all parameters are set via configuration files.

## Configuration
The most important ones are:
* `config.json`: includes all configuration flags and their values
* `peering.json`: includes all connection details to your static peers (neighbors)

> Hornet version 0.5.x targets legacy IOTA 1.0 network. Hornet version 0.6.x targets IOTA 1.5 network aka Chrysalis which is the focus of this documentation

Depending on the installation path you selected, default configuration files may be also part of the installation process and so you may see the following configuration files at your deployment directory:
```bash
config.json
config_chrysalis_testnet.json
config_comnet.json
config_devnet.json
peering.json
profiles.json
```

### Default configuration
By default, Hornet searches for configuration files in the working directory and expects default names, such as `config.json` and `peering.json`.

This behavior can be changed by running Hornet with some altering arguments.

> Please see [config.json](./config.md) and [peering.json](./peering.md) chapters for more information regarding respective configuration files.

Reviewer 2: ```suggestion
Please see the [config.json](./config.md) and [peering.json](./peering.md) chapters for more information regarding respective configuration files.
```

      "passwordHash": "0000000000000000000000000000000000000000000000000000000000000000",
      "passwordSalt": "0000000000000000000000000000000000000000000000000000000000000000"
    }
  }
```
* to enable Dashboard to be reachable from the Internet, it can be changed to `"bindAddress": "0.0.0.0:8081"`

Even if accessible from the Internet, any visitor still needs a valid combination of username and password to access a management section of the Dashboard.

It can be generated using integrated command-line Hornet tools, specifically `pwdhash`:
```bash
hornet tools pwdhash
```

Output example:
```plaintext
Enter a password:
Re-enter your password:
Success!
Your hash: 24c832e35dc542901b90888321dbfc4b1d9617332cbc124709204e6edf7e49f9
Your salt: 6c71f4753f6fb52d7a4bb5471281400c8fef760533f0589026a0e646bc03acd4
```
* `pwdhash` tool provides a combination of `passwordHash` and `passwordSalt` based on your input password
* both values should be copied to the respective configuration values above, specifically the following keys: `passwordHash` and `passwordSalt`

Once Hornet is restarted, then the Dashboard is protected by the given credentials.

## Peer neighbors

> IOTA network is a distributed network in which data is broadcasted among IOTA nodes. IOTA nodes broadcast messages to other IOTA nodes using a gossip protocol. To be able to participate in a network communication, each node has to establish a secure connection to some other nodes in the network - to its peer neighbors - and mutually exchange messages. This is the way how the data is spread within the IOTA network.

Reviewer 2: ```suggestion
The IOTA network is a distributed network in which data is broadcasted among IOTA nodes. IOTA nodes broadcast messages to other IOTA nodes using a gossip protocol. To be able to participate in a network communication, each node has to establish a secure connection to some other nodes in the network - to its peer neighbors - and mutually exchange messages. This is the way how the data is spread within the IOTA network.
```

```

If a node is reachable using DNS name (node01.iota.org) instead of IP address, then the given `multiaddr` would be:
```plaintext
/dns/node01.iota.org/tcp/15600/p2p/12D3KooWHjcCgWPnUEP8wNdbL2fx63Cmosk16xyZ25iUZagxmHb4
```

So node's `multiaddr` can be composed based on generated unique `peerId`, and `bindMultiAddresses` that exists in the `config.json` file under the `p2p` section :
```json
"p2p": {
    "bindMultiAddresses": [
      "/ip4/0.0.0.0/tcp/15600"
    ],
  }
```
* `/ip4/0.0.0.0/tcp/15600`: means it listens to your peer neighbors on IPv4 interface on port 15600 and accepts incoming requests from the Internet
* thanks to used `peer identity`, both parties are able to establish a secure communication channel in which both sides are cryptographically verified and so their mutual identities are confirmed

More information regarding `multiaddr` is also available at [libp2p](https://docs.libp2p.io/concepts/addressing/).

### Adding node peers
Once your node has an unique `multiaddr`, then it can be exchanged with other node owners to establish a mutual peer connection.

*Where to find your future neighbors?*

> Go to the official IOTA Discord server and `#fullnodes` channel, describe your node location (Europe / Germany / Asia, etc.) with your allocated HW resources and ask for some neighbors. Do not publicly disclose your node `multiaddr` to all readers but wait for an individual direct chat.

Reviewer 2: ```suggestion
Go to the official IOTA Discord server and `#fullnodes` channel and describe your node location (Europe / Germany / Asia, etc.) with your allocated HW resources and ask for some neighbors. Do not publicly disclose your node `multiaddr` to all readers but wait for an individual direct chat.
```

Each peer can then be added using the Hornet [Dashboard](#dashboard) (admin section) or [peering.json](./peering.md) file.

A recommended number of peer neighbors is 4-6, since some of them can be offline from time to time.

*Happy peering*

## Configuring HTTP REST API

> One of the [tasks that node is responsible for](../getting_started/nodes_101.md), is exposing HTTP REST API for clients that would like to interacts with the IOTA network, such as crypto wallets, exchanges, IoT devices, etc.

Reviewer 2: ```suggestion
One of the [tasks that node is responsible for](../getting_started/nodes_101.md) is exposing HTTP REST API for clients that would like to interacts with the IOTA network, such as crypto wallets, exchanges, IoT devices, etc.
```
